File: ObsAutoRecord.py
import fnmatch
import os
import threading

import keyboard
import psutil

import ObsUtils
from ObsAutoRecordState import ObsAutoRecordState
from ObsWebSocket import ObsWebSocket

class ObsAutoRecord():
    def __init__(self):
        self.address = ObsUtils.get_address()
        self.apps_to_record = ObsUtils.get_apps_to_record();
        self.default_filename_formatting = '%CCYY-%MM-%DD %hh-%mm-%ss'
        self.interval = 15
        self.obs_web_socket = None
        self.on_state_change = None
        self.timer = None
        self.is_paused = False
        keyboard.add_hotkey('ctrl+alt+end', self.on_pause_key)
        self.start()

    def start(self):
        self.obs_web_socket = ObsWebSocket(
            self.address,
            on_open=self.on_open,
            on_close=self.on_close)

    def on_pause_key(self):
        self.is_paused = not self.is_paused
        self._on_state_change()
        if self.timer is not None:
            self.ping_status()

    def set_on_state_change(self, on_state_change):
        self.on_state_change = on_state_change
        self._on_state_change()

    def on_open(self):
        self._on_state_change()
        self.ping_status()

    def on_close(self):
        self._on_state_change()

    def ping_status(self):
        def start_recording():
            self.obs_web_socket.send("StartRecording")
        def set_filename_formatting(app_name):
            self.obs_web_socket.send(
                "SetFilenameFormatting",
                data={'filename-formatting': app_name + ' - ' + self.default_filename_formatting},
                success_callback=lambda msg: start_recording(),
                error_callback=lambda msg: start_recording())
        def change_folder_back():
            folder = ObsUtils.get_folder()
            if folder is not None:
                self.obs_web_socket.send("SetRecordingFolder", data={'rec-folder': folder})
            self.obs_web_socket.send("SetFilenameFormatting", data={'filename-formatting': self.default_filename_formatting})
        def on_status(msg):
            if 'recording' in msg:
                open_app = self.get_open_app()
                folder = ObsUtils.get_folder()
                if not msg['recording'] and open_app is not None and not self.is_paused:
                    if folder is not None:
                        rec_folder = os.path.join(folder, open_app).replace('\\', '/')
                        self.obs_web_socket.send(
                            "SetRecordingFolder",
                            data={'rec-folder': rec_folder},
                            success_callback=lambda msg: set_filename_formatting(open_app),
                            error_callback=lambda msg: set_filename_formatting(open_app))
                    else:
                        set_filename_formatting(open_app)
                elif msg['recording'] and (open_app is None or self.is_paused):
                    self.obs_web_socket.send(
                        "StopRecording",
                        success_callback=lambda msg: change_folder_back())
        self.obs_web_socket.send("GetStreamingStatus", success_callback=on_status)
        self.timer = threading.Timer(self.interval, self.ping_status)
        self.timer.start()

    def get_open_app(self):
        for proc in psutil.process_iter(attrs=['name']):
            for app in self.apps_to_record:
                if fnmatch.fnmatchcase(proc.name(), app):
                    return ObsUtils.get_app_name_from_process(proc)
        return None

    def _on_state_change(self):
        if self.on_state_change is not None:
            self.on_state_change(self._get_state())

    def _get_state(self):
        if self.is_paused:
            return ObsAutoRecordState.PAUSED
        elif self.obs_web_socket.is_connected():
            return ObsAutoRecordState.CONNECTED
        else:
            return ObsAutoRecordState.DISCONNECTED

File: ObsAutoRecord.cpp
#include "ObsAutoRecord.hpp"

#include <QJsonDocument>
#include <QtCore/QDebug>

QT_USE_NAMESPACE

ObsAutoRecord::ObsAutoRecord(
    const QUrl &url,
    const int &interval,
    const QString &folder,
    QHash<QString, QString> &appsToWatch,
    bool debug,
    QObject *parent) :
        QObject(parent),
        m_obsWebSocket(url, debug, this),
        m_url(url),
        m_interval(interval * 1000),
        m_folder(folder),
        m_appsToWatch(appsToWatch),
        m_debug(debug)
{
    QObject::connect(&m_obsWebSocket, SIGNAL(onResponse(QJsonObject)),
                     this, SLOT(onStatus(QJsonObject)));
    QObject::connect(&m_obsWebSocket, SIGNAL(connected(bool)),
                     this, SLOT(setIsConnected(bool)));
    timer = new QTimer;
    timer->setInterval(m_interval);
    timer->start();
    connect(timer, SIGNAL(timeout()), this, SLOT(pingStatus()));
}

void ObsAutoRecord::setAddress(const QUrl &url)
{
    m_url = url;
    m_obsWebSocket.setAddress(m_url);
}

void ObsAutoRecord::setInterval(const int &interval)
{
    m_interval = interval * 1000;
    timer->setInterval(m_interval);
}

void ObsAutoRecord::setFolder(const QString &folder)
{
    m_folder = folder;
}

void ObsAutoRecord::toggleIsPaused()
{
    m_isPaused = !m_isPaused;
    internalUpdateState();
    pingStatusOverrideTimer();
}

void ObsAutoRecord::pingStatus()
{
    if (m_obsWebSocket.isConnected()) {
        m_msgid++;
        m_obsWebSocket.sendRequest("GetStreamingStatus", m_msgid);
    }
}

void ObsAutoRecord::startRecording()
{
    m_msgid++;
    m_obsWebSocket.sendRequest("StartRecording", m_msgid);
}
        
void ObsAutoRecord::setFilenameFormatting(QString appName, int msgid)
{
    QString filenameFormatting = appName + " - " + defaultFilenameFormatting;
    QJsonObject object
    {
        {"filename-formatting", filenameFormatting}
    };
    m_obsWebSocket.sendRequest("SetFilenameFormatting", msgid, object);
}

void ObsAutoRecord::changeFolderBack()
{
    if (!m_folder.isEmpty()) {
        QJsonObject object
        {
            {"rec-folder", m_folder}
        };
        m_msgid++;
        m_obsWebSocket.sendRequest("SetRecordingFolder", m_msgid, object);
    }
    QJsonObject object2
    {
        {"filename-formatting", defaultFilenameFormatting}
    };
    m_msgid++;
    m_obsWebSocket.sendRequest("SetFilenameFormatting", m_msgid, object2);
}

void ObsAutoRecord::onStatus(QJsonObject msg)
{
    if (msg.contains("message-id")) {
        int msgid = msg.value("message-id").toString().toInt();
        if (idsWaitToRecord.contains(msgid)) {
            idsWaitToRecord.remove(msgid);
            if (idsWaitToRecord.isEmpty()) {
                startRecording();
            }
        }
    }
    if (msg.contains("recording") && idsWaitToRecord.isEmpty()) {
        bool recording = msg.value("recording").toBool();
        QString openApp = ObsUtils::getOpenApp(m_appsToWatch);
        if (!recording && !openApp.isEmpty() && !m_isPaused) {
            if (m_debug) {
                qDebug() << "App found: " << openApp;
            }
            if (!m_folder.isEmpty()) {
                QString recFolder = m_folder;
                if (m_folder.contains("\\")) {
                    // windows paths
                    if (m_folder.endsWith("\\")) {
                        recFolder.append(openApp);
                    } else {
                        recFolder.append("\\" + openApp);
                    }
                } else if (m_folder.contains("/")) {
                    // unix paths
                    if (m_folder.endsWith("/")) {
                        recFolder.append(openApp);
                    } else {
                        recFolder.append("/" + openApp);
                    }
                } else {
                    // not a valid path
                }
                QJsonObject object
                {
                    {"rec-folder", recFolder}
                };
                m_msgid++;
                idsWaitToRecord.insert(m_msgid);
                m_obsWebSocket.sendRequest("SetRecordingFolder", m_msgid, object);
            }
            m_msgid++;
            idsWaitToRecord.insert(m_msgid);
            setFilenameFormatting(openApp, m_msgid);
        } else if (recording && (openApp.isEmpty() || m_isPaused)) {
            m_msgid++;
            m_obsWebSocket.sendRequest("StopRecording", m_msgid);
            changeFolderBack();
        }
    }
}

void ObsAutoRecord::setIsConnected(bool isConnected)
{
    m_isConnected = isConnected;
    internalUpdateState();
    pingStatusOverrideTimer();
}

void ObsAutoRecord::pingStatusOverrideTimer()
{
    if (m_isConnected) {
        timer->setInterval(m_interval);
        pingStatus();
    }
}

void ObsAutoRecord::internalUpdateState()
{
    if (m_isPaused) {
        emit onStateUpdate(ObsAutoRecordState::PAUSED);
    } else {
        if (m_isConnected) {
            emit onStateUpdate(ObsAutoRecordState::CONNECTED);
        } else {
            idsWaitToRecord.clear();
            emit onStateUpdate(ObsAutoRecordState::DISCONNECTED);
        }
    }
}
